ouroboros-consensus/docs/future-ledger-hd.md

```diff
@@ -0,0 +1,219 @@
+
+This document describes the result of the discussion between Ledger and
+Consensus teams on 2022-01-17 about the future steps on UTxO-HD which would not
+make sense to be called this way anymore in future versions and therefore we
+propose Ledger-HD as a replacement.
+
+Below, we outline the tables that are expected to be moved to the disk, their
+dynamics, the main computations that will leave the Ledger and some open
+questions.
+
+This is meant to be just a sketch, details are not worked out yet.
+
+## Scope of Ledger-HD
+
+The plan for Ledger-HD is to move the following tables to the disk:
+
+- The unified map of `rewards`, `delegations`, `pointers` and `deposits`
+
+```haskell
+( ne :: NewEpochState era )
+  & ( nesEs     :: NewEpochState      era  -> EpochState         era  )
+  & ( esLState  :: EpochState         era  -> LedgerState        era  )
+  & ( lsDPState :: LedgerState        era  -> DPState (EraCrypto era) )
+  & ( dpsDState :: DPState (EraCrypto era) -> DState  (EraCrypto era) )
+  & ( dsUnified :: DState  (EraCrypto era) -> UMap    (EraCrypto era) )
+   
+data UMap c = UMap !(Map (Credential 'Staking c) (Trip c)) !(Map Ptr (Credential 'Staking c)) 
+```
+
+> **_NOTE:_** The `deposits` don't seem to be in-place yet? For me it still is a
+> TriMap
+
+- The stake distribution per stake credential
+
+```haskell
+( ne :: NewEpochState era )
+  & ( nesEs           :: NewEpochState era -> EpochState                  era  )
+  & ( esLState        :: EpochState    era -> LedgerState                 era  )
+  & ( lsUTxOState     :: LedgerState   era -> UTxOState                   era  )
+  & ( utxosStakeDistr :: UTxOState     era -> IncrementalStake (EraCrypto era) )
+  
+data IncrementalStake c = IStake
+  { credMap :: !(Map (Credential 'Staking c) Coin)
+  , ptrMap :: !(Map Ptr Coin)
+  }  
+```
+
+- The stake snapshots
+
+```haskell
+( ne :: NewEpochState era )
+  & ( nesEs       :: NewEpochState era -> EpochState           era  )
+  & ( esSnapshots :: EpochState    era -> SnapShots (EraCrypto era) )
+  
+data SnapShots c = SnapShots
+  { ssStakeMark          :: SnapShot c -- Lazy on purpose
+  , ssStakeMarkPoolDistr :: PoolDistr c -- Lazy on purpose
+  , ssStakeSet           :: !(SnapShot c)
+  , ssStakeGo            :: !(SnapShot c)
+  , ssFee                :: !Coin
+  }  
+  
+data SnapShot c = SnapShot
+  { ssStake       :: !(Stake c)
+  , ssDelegations :: !(VMap VB VB (Credential 'Staking c) (KeyHash 'StakePool c))
+  , ssPoolParams  :: !(VMap VB VB (KeyHash 'StakePool c) (PoolParams c))
+  }  
+  
+newtype Stake c = Stake
+  { unStake :: VMap VB VP (Credential 'Staking c) (CompactForm Coin)
+  }  
+  
+newtype PoolDistr c = PoolDistr
+  { unPoolDistr ::
+      Map (KeyHash 'StakePool c) (IndividualPoolStake c)
+  }  
+  
+data IndividualPoolStake c = IndividualPoolStake
+  { individualPoolStake    :: !Rational
+  , individualPoolStakeVrf :: !(Hash c (VerKeyVRF c))
+  }  
+```
+
+> **_TODO:_** All of the maps?
```

Comment:
We probably also want the reward update:

https://github.com/input-output-hk/cardano-ledger/blob/3c4835dc4fe80a19f5714f576c49926d6a58c563/eras/shelley/impl/src/Cardano/Ledger/Shelley/RewardUpdate.hs#L91

the "lens" for it is `NewEpochState -> Maybe RewardUpdate`.